src/cryptoadvance/specter/user.py

```diff
@@ -91,6 +93,26 @@ def __init__(
         # Iterations will need to be increased over time to keep ahead of CPU advances.
         self.encryption_iterations = 390000
 
+        # setting up the notifications system
+        js_notifications = ui_notifications.JSNotifications()
+        webapi_notifications = ui_notifications.WebAPINotifications()
+        self.notification_manager = NotificationManager(
```

Comment:
Ok. I had the manager first instantiated in  `app.specter`. However (of course) it was not differentiating between messages of different users.  I would have to think about how and where to bring the user awareness in.

Reply:
Ideally, this can be done at the endpoint-level. Making the managers non-user-aware is an advantage as then specter can be used independent of a flask-app. From an architectural point of view, that's a signifiant advantage.
However we already broke with that in the user_manager and the service_manager. In the service_manager, i mitigated that issue by assigning the object after the specter-initialisation. Similiar to how the specter-object is assigned to the Flask-App.
I think the mentioned `ServiceEncryptedStorageManager` is mitigating it through it's not so nice Singleton approach.
Maybe we can have a rule like: If it's Flask-aware, assign it to an attribute of Specter after initialisation. If it's not, let it be created in the __init__method of Specter class.
The UserManager is not mitigating that issue right now.

Reply:
I do not understand all, but I experimented on a [branch](https://github.com/cryptoadvance/specter-desktop/compare/master...relativisticelectron:singleton?expand=1) and moved the NotificationManager directly in the Specter class.
- user independent UINotifications are created immediately
- user aware  UINotifications are created later

See here the diff:  https://github.com/cryptoadvance/specter-desktop/compare/master...relativisticelectron:singleton?expand=1

Is that what you had in mind?

Reply:
Passing in the `NotificationManager` imho doesn't improve the situation. Ideally businessObjects don't have many objects. They mostly carry data. Managers on the other hand, shouldn't carry much state and information but rather methods in order to manipulate businessObjects.
If you create a reference from a businessObject towards a manager, no matter whether you've initialized the manager in that businessObject or not, you're complicating things. E.g. testing: Instantiating a User is no longer that easy because you have to pass in a manager and if you pass some Mock of that Manager, you have to avoid using the User in a way which would trigger the Manager.
So, that's why i think that we should avoid making the businessObjects too clever and give them ways to be powerful. They should be passive and manipulated from the outside.

> However (of course) it was not differentiating between messages of different users. 

Well, there are two ways to achieve that:
* You pass in the user for each method of the NotificationManager 
* You let the NotificationManager be aware of flask and let it import `current_user`

I wrote above that it would be cool from an architectural point of view to have managers NOT be flask aware if it's not needed. Because managers are usually instantiated within the specter-object and so, the specter-object gets flask aware as well and you can no longer use the specterobject without flask ( you get: "Outside of ApplicationContext blabla").

So in order to have Managers which are flask aware but still not make specter flask-aware, you can assign the manager after the initialisation of Specter like we do it with the ServiceManager:
https://github.com/relativisticelectron/specter-desktop/blob/master/src/cryptoadvance/specter/server.py#L148-L151

That way it's still comfortable for the calling code to get hold on a reference of that manager via `app.specter`. Hope that clarifies it? I would suggest you do it that way.

Reply:
Thank you for all the feedback and explanation. I will work on it.